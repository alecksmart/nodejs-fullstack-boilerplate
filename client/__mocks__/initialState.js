import { initialState as pointsGame } from '../src/reducers/pointsGame';

export default {
  pointsGame,
};
