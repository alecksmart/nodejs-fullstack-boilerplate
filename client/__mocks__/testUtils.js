import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';

configure({ adapter: new Adapter() });

export const mockStore = configureStore();

export const shallowWithStore = (component, store) => {
  const context = {
    store,
  };
  return shallow(component, { context });
};

export const mountWithStore = (component, store) => {
  const context = {
    store,
  };
  return mount(component, { context });
};
