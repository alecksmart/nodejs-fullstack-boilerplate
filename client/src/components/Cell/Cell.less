.Cell {
  color: #fff;
  border-radius: 5px;
  padding: 8px;

  &--nonclickable {
    background-color: yellowgreen;
  }

  &--clickable {
    background-color: tomato;
  }

  &--active {
    cursor: pointer;
  }

  &--clicked {
    cursor: default;
    background-color: sienna;
  }
}
