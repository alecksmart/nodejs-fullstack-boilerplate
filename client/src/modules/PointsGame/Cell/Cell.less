.Cell {
  color: #fff;
  border-radius: 5px;
  padding: 8px;

  &--nonclickable {
    background-color: rgb(50, 169, 205);
  }

  &--clickable {
    background-color: rgb(195, 251, 255);
  }

  &--active {
    cursor: pointer;
  }

  &--clicked {
    cursor: default;
    background-color: rgb(62, 68, 145);
  }
}
