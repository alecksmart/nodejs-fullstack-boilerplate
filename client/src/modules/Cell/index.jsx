/* eslint-disable react/prop-types */
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import './Cell.less';

class Cell extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      clicked: false,
    };
  }

  onClick = () => {
    const { onPointSelected, x, y, type } = this.props;

    this.setState({ clicked: true });
    onPointSelected({ x, y, type });
  }

  getDisabled = () => {
    const { type, isGameOver } = this.props;
    const { clicked } = this.state;
    return type === '' || clicked || isGameOver;
  }

  render() {
    const { type } = this.props;
    const { clicked } = this.state;

    return (
      <button
        type="button"
        className={`Cell Cell--${type ? 'clickable' : 'nonclickable'} Cell--${clicked ? 'clicked' : 'active'}`}
        disabled={this.getDisabled()}
        onClick={this.onClick}
      >
        {type !== '' && clicked ? type : ' '}
      </button>
    );
  }
}

const mapStateToProps = state => ({
  isGameOver: state.pointsGame.isGameOver,
});

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(Cell);
