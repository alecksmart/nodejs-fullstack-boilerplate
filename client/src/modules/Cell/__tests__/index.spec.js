/* eslint-disable import/no-unresolved */
/* eslint-disable import/order */
import React from 'react';
import { mockStore, shallowWithStore, mountWithStore } from 'testUtils';
import initialState from 'initialState';
import Cell from '../index';

const onPointSelected = jest.fn();

let wrap;
let wrapper;
let props = {
  type: 'A',
  onPointSelected,
};
let context = initialState;
let stateChanged = {};
let propsChanged = {};
let useMount = true;

const store = mockStore(initialState);

describe('>>> Cell snapshot', () => {
  beforeEach(() => {
    wrap = () => {
      context = { ...initialState, ...stateChanged };
      props = { ...props, ...propsChanged };
      return (useMount ? mountWithStore : shallowWithStore)(<Cell store={store} {...props} />, context);
    };
  });

  afterEach(() => {
    wrapper.unmount();
    useMount = true;
    stateChanged = {};
    propsChanged = {};
  });

  it('+++ should capture the snapshot of Cell', () => {
    useMount = false;
    wrapper = wrap();
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('+++ active cell should have propper classes', () => {
    wrapper = wrap();
    const button = wrapper.find('button');

    expect(button).not.toBe(null);
    expect(button.hasClass('Cell--clickable')).toBe(true);
    expect(button.hasClass('Cell--active')).toBe(true);
  });

  it('+++ simple cell should have propper classes', () => {
    propsChanged = { type: false };
    wrapper = wrap();
    const button = wrapper.find('button');

    expect(button).not.toBe(null);
    expect(button.hasClass('Cell--nonclickable')).toBe(true);
    expect(button.hasClass('Cell--active')).toBe(true);
  });

  it('+++ clicks on button should take correct effects', () => {
    wrapper = wrap();
    const button = wrapper.find('button');
    button.simulate('click');
    expect(onPointSelected).toBeCalledTimes(1);
  });
});
