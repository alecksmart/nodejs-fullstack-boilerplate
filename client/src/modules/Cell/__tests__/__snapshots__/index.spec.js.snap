// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Cell snapshot +++ should capture the snapshot of Cell 1`] = `
"<ContextProvider value={{...}}>
  <Cell store={{...}} type=\\"A\\" onPointSelected={[Function: mockConstructor]} isGameOver={false} />
</ContextProvider>"
`;
